functionalDB/chapter.md

```diff
@@ -0,0 +1,1097 @@
+
+````clojure
+(defn- update-attr-value [attr value operation]
+   (cond
+      (single? attr)    (assoc attr :value #{value})
+    ; now we're talking about an attribute of multiple values
+      (= :db/reset-to operation)  (assoc attr :value value)
+      (= :db/add operation) (assoc attr :value (CS/union (:value attr) value))
+      (= :db/remove operation) (assoc attr :value (CS/difference (:value attr) value))))
+````
+All that remains is to remove the old value from the indexes and to add the new one to them, and then construct to the new layer with all of our updated components. Luckily, we can leverage the code we wrote for adding and removing entities to do this!
+
+### Transactions
+
+All the operations described before (add / remove / edit) do a single operation on a single entity. The return value from them is the database as it was before the operation  topped with an additional layer. Yet, a key requirement of any database is the ability to perform a transaction that includes several operations, possibly on several elements, while:
+
+* Viewed from the outside as a single (i.e., atomic) operation, meaning either all operations succeed or all operations fails
+* Keep the database in a valid state at any given time 
+* Keep the update operations isolated from the outside at any point in time until they are completed. 
+
+The way to answer these requirements is by using a mechanism that takes as an input input the database and a set of operations to be performed, and produces as an output a database whose state reflects the given changes. More than that, all the changes should be seen as an addition of a single layer. 
+
+Performing a transactional operation is done in Clojure using the *Atom* element (this is why we wrap a database with such an element). This element allow us to invoke a single function on it, and all changes to it happen in an atomic and isolated manner. 
+
+As for the function that is to be executed, this function needs to perform all the operations that the user requested, while eventually adding one layer to the database.  To do so we would need to overcome the fact that each of the lifecycle functions adds a layer reflecting the change it inflicts. 
+
+Here we resort to the insight that changes on different attributes (that can be part of either the same or different entities) accumulate when stacking layers, thus a top layer holds all the changes that built the layers below it. Therefore, the solution is to execute each of the user’s operations one after another, each creating a new layer. While doing so, we do not update the databases’s timestamp field. When the last layer is completed, we take only that top layer and place it on the initial database  (leaving all the intermediate layers to pine for the fjords) and only then update the database’s timestamp. All this is happening in the *transact-on-db* function
+
+````clojure
+(defn transact-on-db [initial-db txs]

```

Comment:
I think `txs` is a confusing name for this parameter. We're not really passing in transactions, are we? It's more like a list of operations that we're executing as a single transaction. Maybe `ops` or `updates`?